### ISSUE TITLE
Problem with JWT token authentication in the pilot & envoy

### ISSUE THREAD

**Issue by piotrmsc** [istio/istio#10547]
**Describe the bug**
When you use `End User Authentication Policy` to secure a service pilot is refreshing **JWKS** cache each hour. Once an IDP rotates new keys and you fetch a new token signed with a new key, envoy rejects it due to lack of JWKS configuration. Request results in 401 with **KID_ALG_UNMATCH** error in the envoy logs. 

Looking at the implementation of the filter (https://github.com/istio/proxy/blob/master/src/envoy/http/jwt_auth/jwt.cc#L430) it looks like
the cache is the only 'source of truth' for an envoy regarding JWT authentication. The pilot is responsible for pushing new JWKS configuration to the envoys. In such case, in the worst case scenario, there is a 1-hour gap until an envoy will receive new JWKS, therefore, will reject new tokens for this period of time.
Shouldn't envoy make a call for new JWKS if it detects missing piece in the cache despite pilot job?
Restarting pod with running envoy helps after that new tokens are properly authenticated.

Bellow, I attach debug logs from an envoy injected to an application and pilot as well. I can see that the pilot is refreshing keys each hour.

**Expected behavior**
If a request is made to the service with token signed by a new key, envoy is able to fetch new JWKS configuration in order to authenticate the token.

**Steps to reproduce the bug**
1. Install istio
2. Install dex and set key rotation period to for example to 10 minutes (to speed up problem reproduction)
3. Secure sample service with EndUser Policy pointing to dex
4. Fetch a token after key rotation
5. Make a request with a new token to the service

In my case I use Dex as an IDP. To eliminate the problem in Dex itself, I have verified that new keys are available under public kyes endpoint.

**Version**
Istio
Version: 1.0.5

Kubectl
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-28T15:18:13Z", GoVersion:"go1.11", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-04-10T12:46:31Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"linux/amd64"}

**Installation**
{{ Please describe how Istio was installed }}
Via Helm chart with security enabled

**Environment**
{{ Which environment, cloud vendor, OS, etc are you using? }}
Minikube & cluster env like GKE

**Cluster state**
{{ If you're running on Kubernetes, consider [following the
instructions](http://istio.io/help/bugs/#generating-a-cluster-state-archive)
to generate "istio-dump.tar.gz", then attach it here by dragging and dropping
the file onto this issue. }}
[envoy.log](https://github.com/istio/istio/files/2689877/envoy.log)
[pilot.log](https://github.com/istio/istio/files/2689878/pilot.log)

**Comment by liminw** [comment]
To rotate keys, you need to strictly follow the lifecycle of a key. In the following steps, from step 2 to step 3 is the duration you can use a key to sign tokens. 

1. Add a new key (k1) to the key server. At this time, you cannot use k1 to sign any token yet. Wait for the duration of cache refresh, and k1 is fetched by Pilot. 
2. Start using k1 to sign tokens. 
3. Stop using k1 to sign tokens. At this time, you have started to retire k1. 
4. Wait until all tokens signed by k1 has expired (longest token expiration time). After that, you can delete k1 from the key server.

**Comment by piotrmsc** [comment]
Hey @liminw  thx for the reply :) 

I have a question regarding what you have mentioned, according to https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys

> " The verifier knows to go back to the jwks_uri location to re-retrieve the keys when it sees an unfamiliar kid value."

in this case the verifier is an envoy not pilot, however, envoy has config based on what pilot provides. So with unmatched kid envoy should re-retrieve the keys which currently is missing.

The draft does not mention that the issuer must keep keys as long as id token is valid, it mentions it should keep old/expired keys for a reasonable time. 
In my opinion the JWKS handling is broken, as envoy relies only on pilot here. Envoy should be able to re-retrive the JWKS by itself.
What's your thoughts? :)

**Comment by yangminzhu** [comment]
Previously we do let Envoy to fetch the public key, we switched to let Pilot to fetch and distribute the public key later. There are a couple of reasons for this change, 1) If you have a lot envoy in the cluster, each of them will trigger a unique request to fetch the public key, 2) It implies we have to let the Envoy to be able to access the external Internet but this is often not allowed in some environments. So we move this to the control plane (i.e. Pilot).

We're making the refresh interval configurable in Pilot: [PR](https://github.com/istio/istio/pull/11424). Would this mediate the issue you're facing here? 

/cc @diemtvu @pitlv2109

**Comment by piotrmsc** [comment]
It will help but unfortunately will not solve the issue. Even if you set refresh time to 1 minute it will increase load on pilot and still for this period of time tokens will be rejected. I understand the problem with access to the internet of the Envoy, however, pilot needs this access as well so if the cloud policy blocks such calls it will not be possible to use AuthN policy at all cause even pilot will not be able to make well-known request. However, I understand that it's better to allow a single component to have access to the external world rather then N components...

But what if the model would change? So pilot would be still responsible for providing keys to the envoys, however, envoy as a verifier, when detects missing keys in cache could ask pilot for new config, despite periodic refresh time? This way you would have system running all the time able to authenticate new ID tokens regardless cache refreshing.

When do you plan to release this PR with configurable refresh time ?:)

**Comment by yangminzhu** [comment]
@pitlv2109 is working to make the refresh time configurable.

The model you described is very similar to the [SDS (Secret Discovery Service)](https://preliminary.istio.io/docs/concepts/security/#node-agent-in-kubernetes) we added in 1.1, currently it's only used to fetch the certificate/key for Envoy. Technically It's possible to extend it to also fetch the JWT public key, but it's not a trivial change. @diemtvu @quanjielin WDYT?

**Comment by zhaorendong** [comment]
Can you provide an interface or API for pilot so that we can call the API to update the JWk cache in pilot?

**Comment by piotrmsc** [comment]
I don't think it should stale... the issue still exists and mentioned workaround pr (https://github.com/istio/istio/pull/11424)  is closed with "will be done in the future" message.   
 @quanjielin @diemtvu  it looks like you're at kubecon Barcelona now, maybe we could discuss this?

**Comment by pitlv2109** [comment]
We will address this issue in the near future. We're avoiding adding more flags to Istio control plane components and will probably introduce a new way to handle this.

**Comment by jerriclynsjohn** [comment]
Looking forward for this, right now I'm still handling the verification in my service itself, this is really an essential add on to end-user authentication and authorization in Istio.

**Comment by JimmyCYJ** [comment]
@liminw @yangminzhu are there any updates on this?

**Comment by StymiedSloth** [comment]
Hello
> We will address this issue in the near future. We're avoiding adding more flags to Istio control plane components and will probably introduce a new way to handle this.

Any updates on this? 
Thanks

**Comment by mhite** [comment]
We are hitting this issue when trying to integrate with https://www.gluu.org/. Apparently they have no grace period for key rotation in their implementation. Once they generate a new key, it is their expectation that they can immediately use it.

**Comment by mhite** [comment]
Could there conceivably ever be an option of choosing for Pilot to send jwksUri to envoy rather than jwks only? (Assuming Envoy supports refreshing against jwksUri when unrecognized kid arrives and envoy has network access to jwksUri.)

Could that be a simpler alternative than waiting for an SDS like implementation of key rotation for jwt?

**Comment by ayush-jain1** [comment]
@piotrmsc @yangminzhu  We have same requirement, in case of kid miss in JWKS cache, we would like to refresh the cache as possibly there is a key rotation, 

Has this change already implemented in istio ? Also is JWKS cache TTL configurable with ISTIO ?

**Comment by asador** [comment]
> To rotate keys, you need to strictly follow the lifecycle of a key. In the following steps, from step 2 to step 3 is the duration you can use a key to sign tokens.
> 
>     1. Add a new key (k1) to the key server. At this time, you cannot use k1 to sign any token yet. Wait for the duration of cache refresh, and k1 is fetched by Pilot.
> 
I wish step 1 was respected by token providers. What you are says is not defined in OpenID standard.

**Comment by bmd007** [comment]
> https://www.gluu.org/

key providers don't really care about caching strategies of consumers. They are only bound to keep the public key longer than the private key. The rest is on the consumers to not shoot themselves in the foot.

**Comment by bmd007** [comment]
> > To rotate keys, you need to strictly follow the lifecycle of a key. In the following steps, from step 2 to step 3 is the duration you can use a key to sign tokens.
> > ```
> > 1. Add a new key (k1) to the key server. At this time, you cannot use k1 to sign any token yet. Wait for the duration of cache refresh, and k1 is fetched by Pilot.
> > ```
> 
> @liminw I wish step 1 was respected by token providers. What you are says is not defined in OpenID standard.

key providers don't really care about caching strategies of consumers. They are only bound to keep the public key longer than the private key. The rest is on the consumers to not shoot themselves in the foot. Which in this case, imo, we are by not allowing a refetch of jwks when key(Id) is missing from jwks.

**Comment by bmd007** [comment]
any updates? is it still not possible to ask pilot to reftech jwkSet when key(id) missing from currently fetched jwkSet?

**Comment by martin2176** [comment]
I see that the new Istiod Pilot support this env variable.
It supports "Envoy" . So I am assuming instead of Istiod fetching jwks, caching it and sending it out to envoys, envoys fetch jwks directly.
However it is not documented how envoy fetches jwks or atleast I was not able to find any info on the logic.
Envoy also maintain a cache of jwsk AFAIK. But does envoy fetch JWKS from JWKURI when it encounter a new KID not found in cache.
That is really the bottom line .

PILOT_JWT_ENABLE_REMOTE_JWKS
Mode of fetching JWKs from JwksUri in RequestAuthentication. Supported value: istiod, false, hybrid, true, envoy. The client fetching JWKs is as following: istiod/false - Istiod; hybrid/true - Envoy and fallback to Istiod if JWKs server is external; envoy - Envoy.

**Comment by ianonavy** [comment]
> I see that the new Istiod Pilot support this env variable. It supports "Envoy" . So I am assuming instead of Istiod fetching jwks, caching it and sending it out to envoys, envoys fetch jwks directly. However it is not documented how envoy fetches jwks or atleast I was not able to find any info on the logic. Envoy also maintain a cache of jwsk AFAIK. But does envoy fetch JWKS from JWKURI when it encounter a new KID not found in cache. That is really the bottom line .
> 
> PILOT_JWT_ENABLE_REMOTE_JWKS Mode of fetching JWKs from JwksUri in RequestAuthentication. Supported value: istiod, false, hybrid, true, envoy. The client fetching JWKs is as following: istiod/false - Istiod; hybrid/true - Envoy and fallback to Istiod if JWKs server is external; envoy - Envoy.

@martin2176 I just tested this on Istio 1.18.0 (before 1.18 Envoy is only supported for in-mesh JWKs servers), and unfortunately it did not work. The Envoy proxy also has a cache ([5 min as configured by Istio](https://github.com/istio/istio/blob/5f5d657c72d30a97cae97938de3a6831583e9f15/pilot/pkg/security/authn/v1beta1/policy_applier.go#L270) but [the default is 10 min](https://github.com/envoyproxy/envoy/blob/5c821febbbba530315295a3c3dd3481364f89890/api/envoy/extensions/filters/http/jwt_authn/v3/config.proto#L356-L357)). It would appear that the [Envoy team had the same idea to invalidate the cache on new kid, but it's still a TODO](https://github.com/envoyproxy/envoy/blob/5c821febbbba530315295a3c3dd3481364f89890/source/extensions/filters/http/jwt_authn/authenticator.cc#L251-L257).

**Comment by martin2176** [comment]
I opened a enhancement request
 https://github.com/istio/istio/issues/49044#issuecomment-1914030728
 
 I did some rudimentary testing. 
 I am using the default setitng "PILOT_JWT_ENABLE_REMOTE_JWKS istiod" where Istiod is pulling down JWKS and distributing to envoys.
 It looks like when Istiod (every PILOT_JWT_PUB_KEY_REFRESH_INTERVAL) , if it finds a new JWKS KID, it will push the jwks to all the envoys. It only pushes the JWKS only if it finds an updated JWKS. As far as I see in istiod logs, if there is no change in JWKS, there is no pushing out JWKS to envoys.
 
 Envoy (though have a cache) , when it receives an updated JWKS (with new KID) envoy immediately starts to use it for verification. Envoy cache is not coming into play there. it doesnt wait an additional 5 or 10minutes.
  
 So technically , purely based on my testing:
 If Istiod can detect a unknown KID situation (based on status in envoy failed_status_in_metadata) and refresh JWKS immediately, technically only few inflight auth requests should fail.
